File: 0-video-ingestion-and-transcoding/0-typescript/src/transcode/transcode.controller.ts
import { Body, Controller, Get, HttpCode, HttpStatus, NotFoundException, Param, Post } from "@nestjs/common"
import { TranscodeService } from "./transcode.service"

/** Request body for POST /api/videos/transcode. */
class CreateTranscodeDto {
    inputPath: string
}

@Controller("api/videos/transcode")
export class TranscodeController {
    constructor(private readonly transcodeService: TranscodeService) {}

    @Post()
    @HttpCode(HttpStatus.ACCEPTED) // 202: accepted, FFmpeg runs in the background.
    async create(@Body() dto: CreateTranscodeDto) {
        const job = await this.transcodeService.createJob(dto.inputPath)
        return {
            id: job.id,
            inputPath: job.inputPath,
            outputPath: job.outputPath ?? null,
            status: job.status,
            errorMessage: job.errorMessage ?? null,
            createdAt: job.createdAt,
            updatedAt: job.updatedAt,
        }
    }

    @Get(":id")
    async get(@Param("id") id: string) {
        const job = await this.transcodeService.findOne(id)
        if (!job) throw new NotFoundException("Job not found")
        const outputUrl = await this.transcodeService.presignedFor(job)
        return {
            id: job.id,
            inputPath: job.inputPath,
            outputPath: job.outputPath ?? null,
            status: job.status,
            errorMessage: job.errorMessage ?? null,
            outputUrl,
            createdAt: job.createdAt,
            updatedAt: job.updatedAt,
        }
    }
}

File: 0-video-ingestion-and-transcoding/1-java/src/main/java/com/starci/transcode/TranscodeController.java
package com.starci.transcode;

import com.starci.transcode.TranscodeDtos.TranscodeJobDto;
import com.starci.transcode.TranscodeDtos.TranscodeRequest;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;

@RestController
@RequestMapping("/api/videos/transcode")
public class TranscodeController {

    private final TranscodeService transcodeService;
    private final TranscodeJobRepository jobRepository;
    private final StorageService storage;

    public TranscodeController(TranscodeService transcodeService,
                               TranscodeJobRepository jobRepository,
                               StorageService storage) {
        this.transcodeService = transcodeService;
        this.jobRepository = jobRepository;
        this.storage = storage;
    }

    @PostMapping
    public ResponseEntity<TranscodeJobDto> createJob(@RequestBody TranscodeRequest req) {
        TranscodeJob job = transcodeService.createJob(req.inputPath());
        // 202 Accepted — the job is PENDING and FFmpeg runs in the background.
        return ResponseEntity.status(HttpStatus.ACCEPTED).body(TranscodeJobDto.from(job, null));
    }

    @GetMapping("/{id}")
    public ResponseEntity<TranscodeJobDto> getJob(@PathVariable String id) {
        TranscodeJob job = jobRepository.findById(id)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Job not found"));

        String outputUrl = null;
        if (job.getStatus() == JobStatus.COMPLETED && job.getOutputPath() != null) {
            outputUrl = storage.presignedGetUrl(job.getOutputPath());
        }
        return ResponseEntity.ok(TranscodeJobDto.from(job, outputUrl));
    }
}
